src/hotspot/share/opto/inlinetypenode.cpp

```diff
@@ -584,9 +590,15 @@ InlineTypeNode* InlineTypeNode::buffer(GraphKit* kit, bool safe_for_replace) {
       kit->kill_dead_locals();
       Node* klass_node = kit->makecon(TypeKlassPtr::make(vk));
       Node* alloc_oop  = kit->new_instance(klass_node, nullptr, nullptr, /* deoptimize_on_exception */ true, this);
-      // No need to initialize a larval buffer, we make sure that the oop can not escape
-      if (!is_larval()) {
-        // Larval will be initialized later
+
+      // Larval?
+      // - must_init set: We are about to call an abstract value class constructor or the Object constructor which is
+      //                  not inlined. It is therefore escaping and we must initialize the buffer because we have not
+      //                  done this, yet, for larvals (see second case).
+      // - Other larvals: We do not need to initialize the buffer because a larval could still be updated which will
+      //                  create a new buffer. Once the larval escapes, we will initialize the buffer (must_init set).
+      assert(!must_init || is_larval(), "must_init should only be set for larval");
+      if (!is_larval() || must_init) {
```

Comment:
This is confusing: Shouldn't the default of `must_init` be true because that's the common case? And then we can assert here that it can only ever be false if `is_larval()` is true.